### DISCUSSION TITLE
Pick better name for toCheckedValue and fromCheckedValue

### DISCUSSION THREAD

**Discussion by reinco** [ndsev/zserio-cpp17#61] [Ideas]
Instead of using the cryptic names why do we not just make several template functions, called 'zserio::cast' ?

Example on how the API could look like (no actual numerical range checking is implemented):
```c++
template <typename T>
class Number
{
public:   
    using BaseType = T;

    Number(T t) { m_value = t; }

    BaseType value() { return m_value; }
private:   
   BaseType m_value;
};

template <typename T>
T cast(typename T::BaseType value)
{
    return T(value); 
}

template <typename T, typename U = Number<T>>
T cast(U value)
{
    return value.value();
}

int main()
{
    auto v = cast<Number<int>>(4);
    auto w = cast<int>(v);
    return 0;
}
```

**Comment by mikir** [comment]
I have two notes regarding this solution:

1. Naming is difficult. From my point of view, `cast` is cryptic name as well because it is not immediately obvious whether the function performs range checking before casting, which could lead to unexpected behavior. It might be that something like `checked_cast` will be better.
2. The cast from wrapper type to the native type is a little bit misleading because wrapper type already knows very well the native type. So, it might be better to implement something like this:

   ```
   template <typename T>
   T cast(Number<T> value)
   {
       return value.value();
   }
   ```
   This allows to write something like
   ```
   int w = cast<int>(v);
   ```
   or
   ```
   int w = cast(v);
   ```
   but not something like
   ```
   int w = cast<unsigned>(v);
   ```

   If `cast` without `<int>` looks strange, we can disable it. But then, users will have to specify something which might be redundant.
